haystack/retriever/dense.py

```diff
@@ -120,57 +99,142 @@ def embed_queries(self, texts: List[str]) -> List[np.array]:
-        if len(titles) != len(texts):
-            titles = None  # type: ignore
-
-        result = self._generate_batch_predictions(texts=texts, titles=titles, model=self.passage_encoder,
-                                                  tensorizer=self.tensorizer, batch_size=self.batch_size)
+            titles = [d.meta["name"] if d.meta and "name" in d.meta else "" for d in docs]
+
+        result = self._generate_batch_predictions(texts=texts, titles=titles,
+                                                  model=self.passage_encoder,
+                                                  tokenizer=self.passage_tokenizer,
+                                                  batch_size=self.batch_size)
         return result
 
+    def _normalize_query(self, query: str) -> str:
+        if query[-1] == '?':
+            query = query[:-1]
+        return query
+
+    def _tensorizer(self, tokenizer: Union[DPRQuestionEncoderTokenizer, DPRContextEncoderTokenizer],
+                    text: List[str],
+                    title: Optional[List[str]] = None,
+                    add_special_tokens: bool = True):
+        """
+        Creates tensors from text sequences
+        :Example:
+            >>> ctx_tokenizer = DPRContextEncoderTokenizer.from_pretrained()
+            >>> dpr_object._tensorizer(tokenizer=ctx_tokenizer, text=passages, title=titles)
+
+        :param tokenizer: An instance of DPRQuestionEncoderTokenizer or DPRContextEncoderTokenizer.
+        :param text: list of text sequences to be tokenized
+        :param title: optional list of titles assoicated with each text sequence
+        :param add_special_tokens: boolean for encoding of special tokens in each sequence
+
+        Returns:
+                token_ids: list of token ids from vocabulary
+                token_type_ids: list of token type ids
+                attention_mask: list of indices specifying which tokens should be attended to by the encoder
+        """
+
+        # combine titles with passages only if some titles are present with passages
+        if self.embed_title and title:
+            final_text = [tuple((title_, text_)) for title_, text_ in zip(title, text)] #type: Union[List[Tuple[str, ...]], List[str]]
+        else:
+            final_text = text
+        out = tokenizer.batch_encode_plus(final_text, add_special_tokens=add_special_tokens, truncation=True,
+                                              max_length=self.max_seq_len,
+                                              pad_to_max_length=True)
+
+        token_ids = torch.tensor(out['input_ids']).to(self.device)
+        token_type_ids = torch.tensor(out['token_type_ids']).to(self.device)
+        attention_mask = torch.tensor(out['attention_mask']).to(self.device)
+        return token_ids, token_type_ids, attention_mask
+
+    def _remove_sep_tok_from_untitled_passages(self, titles, ctx_ids_batch, ctx_attn_mask):
+        """
+        removes [SEP] token from untitled samples in batch
```

Comment:
Add a small explanation here why we give this option + maybe a small example

Reply:
explanation and examples added